### PR TITLE
Deal with special case for Array[Unit] in Array.{copyOf,copyAs}

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

**Comment by retronym** [comment]
Thanks to @mkeskells for spotting that we were filling `Arrry[Unit]` with `null`-s.

**Comment by sjrd** [comment]
That's not false, but fixing this seems like a drop in the ocean. There are many other cases where constructing an array of `Unit`s fills it with `null`. `new Array[Unit](n)`, for starters:
```scala
$ ~/opt/scala-2.13.0-M4/bin/scala
Welcome to Scala 2.13.0-M4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_40).
Type in expressions for evaluation. Or try :help.

scala> val a = new Array[Unit](5)
a: Array[Unit] = Array((), (), (), (), ()) // This is a lie

scala> a.asInstanceOf[Array[scala.runtime.BoxedUnit]](0)
res0: scala.runtime.BoxedUnit = null
```
The lie comes from [`ArraySeq.ofUnit.apply`](https://github.com/scala/scala/blob/69f31a254ad4532f19aa2a7460082a559b6d2347/src/library/scala/collection/mutable/ArraySeq.scala#L262) (or `WrappedArray.ofUnit.apply` in earlier versions), which returns a hard `Unit`, and therefore the result of `array(index)` is actually dropped, and a new `BoxedUnit.UNIT` is created in the *bridge* `apply(int)Object` generated to implement the inherited `def apply(index: Int): T`.

File: src/library/scala/Array.scala [hunk 1]
```diff
@@ -122,7 +122,9 @@ object Array {
     * @see `java.util.Arrays#copyOf`
     */
   def copyOf[A](original: Array[A], newLength: Int): Array[A] = (original match {
-    case x: Array[AnyRef]     => java.util.Arrays.copyOf(x, newLength)
+    case x: Array[AnyRef]     =>
+      if (x.getClass.getComponentType == classOf[BoxedUnit]) newUnitArray(newLength).asInstanceOf[Array[A]]
```
Comment (szeiger) [inline src/library/scala/Array.scala]: Any reason not to use the usual match `case x: Array[BoxedUnit]`?

**Review by szeiger — APPROVED** [review]

File: src/library/scala/Array.scala (hunk 1 above)
Comment (retronym) [reply src/library/scala/Array.scala]: I've changed to use that. The only tricky part is that it needs to precede the match on `_: Array[AnyRef]` due to array covariance in the VM.

**Comment by xuwei-k** [comment]
BTW https://github.com/xuwei-k/scala-1/commit/a9761ba2ca0c687b8690416f3b733c82d3f5ad9e 🤔

**Comment by szeiger** [comment]
@xuwei-k The JVM doesn't have `void` arrays, therefore `Array[Unit]` is really `Array[BoxedUnit]`. The pattern matcher would have to explicitly check for that case when matching on `Array[AnyRef]`. It's not exactly a common case, favoring correctness over performance seems reasonable. I suggest filing a bug.

**Pull request merged**